### PR TITLE
Campaign Module Migration Utility Added For Respark

### PR DESCRIPTION
**Pull request by MayureshRelfor**

### CODE REVIEW

File: src/main/java/com/relfor/pcs/PcsUtilityApplication.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    private ApplicationContext applicationContext;
+
+    public static void main(String[] args) {
+        SpringApplication.run(PcsUtilityApplication.class, args);
+    }
+
+    @PostConstruct
+    public void executeMigration() {
+        String[] args = applicationArguments.getSourceArgs();
+        if (args.length == 1) {
```
Comment (DevourinAdmin) [inline src/main/java/com/relfor/pcs/PcsUtilityApplication.java:34]: validate the argument. I can garbage.
The "args" in the main method, why is that not utilised?

File: src/main/java/com/relfor/pcs/PcsUtilityApplication.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            Utilities utility = utilityFactory.getUtility(args[0]);
+            log.debug("Migration Utility Started - " + args[0]);
+            utility.migrateDataFromMongoToMySQL();
```
Comment (DevourinAdmin) [inline src/main/java/com/relfor/pcs/PcsUtilityApplication.java:37]: Why a specific method called here? We discussed about having a generic performMigration. Will you keep changing this line everytime you want to run a diff migration utility?

File: src/main/java/com/relfor/pcs/PcsUtilityApplication.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            log.debug("Migration Utility Ended - " + args[0]);
+        } else {
+            log.error("Pass 1 Utility Type in Arguments");
+            throw new IllegalStateException("Pass 1 Utility Type in Arguments");
+        }
+        shutdown();//Manually shut down pcs-utility application
+    }
+
+    private void shutdown() {
+        int exitCode = SpringApplication.exit(applicationContext, () -> 0);
```
Comment (DevourinAdmin) [inline src/main/java/com/relfor/pcs/PcsUtilityApplication.java:47]: who calls shutdown method?

File: src/main/java/com/relfor/pcs/PcsUtilityConfig.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        return new WebMvcConfigurer() {
+            @Override
+            public void addCorsMappings(CorsRegistry registry) {
+                registry.addMapping("/**").allowedOrigins("*").allowedMethods("POST", "PUT", "GET");
+                WebMvcConfigurer.super.addCorsMappings(registry);
+            }
+        };
+    }
+
+    @Bean
```
Comment (DevourinAdmin) [inline src/main/java/com/relfor/pcs/PcsUtilityConfig.java:42]: why do you need this web stuff?

File: src/main/java/com/relfor/pcs/util/CampaignModuleMigrationUtility.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        this.templatesRepository = templatesRepository;
+        this.mongoPromotionsRepository = mongoPromotionsRepository;
+        this.designStateRepository = designStateRepository;
+        this.mongoTemplatesRepository = mongoTemplatesRepository;
+        this.apiHelper = apiHelper;
+    }
+
+    public void migrateDataFromMongoToMySQL() {
+        // GET ONLY ACTIVE TENANT FROM PCS
+        List<Long> tenantIdList = apiHelper.getAllActiveTenantIdList();
```
Comment (DevourinAdmin) [inline src/main/java/com/relfor/pcs/util/CampaignModuleMigrationUtility.java:39]: Ensure that the data has not been migrated before.

File: src/main/resources/application.properties [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+springfox.documentation.enabled=false
+spring.mvc.pathmatch.matching-strategy=ANT_PATH_MATCHER
+spring.cache.type=simple
+#URL
+pcs_url=https://localhost:8081
+pcs_catalog_url=https://localhost:8082
+pcs_txn_url=https://localhost:8084
+pcs_guest_url=https://localhost:8083
+pcs_inventory_url=https://localhost:8086
+pcs_utility_url=https://localhost:8091
```
Comment (DevourinAdmin) [inline src/main/resources/application.properties:34]: this is not a service.

File: src/main/resources/application.properties [hunk 7]
```diff
@@ -0,0 +1,35 @@
+spring.application.name=pcs-utility
+server.servlet.context-path=/pcs-utility
+spring.profiles.active=default
+resource.bundle.basename=i18n.messages
+server.port=8091
```
Comment (DevourinAdmin) [inline src/main/resources/application.properties:5]: not required

File: src/main/java/com/relfor/pcs/PcsUtilityApplication.java (hunk 2 above)
Comment (MayureshRelfor) [reply src/main/java/com/relfor/pcs/PcsUtilityApplication.java:37]: I have now renamed it to performMigration.

File: src/main/java/com/relfor/pcs/PcsUtilityApplication.java (hunk 3 above)
Comment (MayureshRelfor) [reply src/main/java/com/relfor/pcs/PcsUtilityApplication.java:47]: The shutdown method will be called after the completion of executeMigration(), whether it succeeds or fails.

File: src/main/resources/application.properties (hunk 6 above)
Comment (MayureshRelfor) [reply src/main/resources/application.properties:34]: Removed

File: src/main/resources/application.properties (hunk 7 above)
Comment (MayureshRelfor) [reply src/main/resources/application.properties:5]: Removed

File: src/main/java/com/relfor/pcs/PcsUtilityApplication.java (hunk 3 above)
Comment (MayureshRelfor) [reply src/main/java/com/relfor/pcs/PcsUtilityApplication.java:47]: Added try catch finally.

File: src/main/java/com/relfor/pcs/PcsUtilityApplication.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Autowired
+    public PcsUtilityApplication(UtilityFactory utilityFactory, ApplicationContext applicationContext) {
+        this.utilityFactory = utilityFactory;
+        this.applicationContext = applicationContext;
+    }
+
+    public static void main(String[] args) {
+        utilityName = args.length > 0 ? args[0] : null;
+        SpringApplication.run(PcsUtilityApplication.class, args);
```
Comment (DevourinAdmin) [inline src/main/java/com/relfor/pcs/PcsUtilityApplication.java:31]: Add some SOPs for the console.
Utility to migrate started....
followed by various SOPs that decribe significant milestones during the execution of the utility
Lastly Migration successfully completed. .
...Terminating
These SOPs on the console will keep the user engaged and provide him the progress.

Comment (MayureshRelfor) [reply src/main/java/com/relfor/pcs/PcsUtilityApplication.java:31]: The same SOPs are added in executeMigration() and performMigration() method

**Review by DevourinAdmin — APPROVED** [review]
Please test with real data, we need to check how much time the utility takes. Also after the migration test how much time it takes to fetch/create/modify. Note current timings for theese operations prior to migration and then compare those with post migration.

**Pull request merged**